drivers/modem/hl7800.c

```diff
@@ -0,0 +1,4887 @@
+		CMD_HANDLER("+CEER: ", sockerror),
+		CMD_HANDLER("+KTCPCFG: ", sockcreate),
+		CMD_HANDLER("+KUDPCFG: ", sockcreate),
+		CMD_HANDLER(CONNECT_STRING, connect),
+		CMD_HANDLER("NO CARRIER", sockerror),
+
+		/* UNSOLICITED SOCKET RESPONSES */
+		CMD_HANDLER("+KTCP_IND: ", sock_ind),
+		CMD_HANDLER("+KUDP_IND: ", sock_ind),
+		CMD_HANDLER("+KTCP_NOTIF: ", sock_notif),
+		CMD_HANDLER("+KUDP_NOTIF: ", sock_notif),
+		CMD_HANDLER("+KTCP_DATA: ", sockdataind),
+		CMD_HANDLER("+KUDP_DATA: ", sockdataind),
+
+		/* FIRMWARE UPDATE RESPONSES */
+		CMD_HANDLER("+WDSI: ", device_service_ind),
+	};
+
+	while (true) {
+		/* wait for incoming data */
+		k_sem_take(&ictx.mdm_ctx.rx_sem, K_FOREVER);
+
+		hl7800_read_rx(&rx_buf);
+		/* If an external module hasn't locked the command processor,
+		 * then do so now.
+		 */
+		if (!hl7800_RX_locked()) {
+			hl7800_RX_lock();
+			unlock = true;
+		} else {
+			unlock = false;
+		}
+
+		while (rx_buf) {
+			removeLineFromBuf = true;
+			cmd_handled = false;
+
+#ifdef CONFIG_MODEM_HL7800_FW_UPDATE
+			if ((ictx.fw_update_state == HL7800_FOTA_START) ||
+			    (ictx.fw_update_state == HL7800_FOTA_WIP) ||
+			    (ictx.fw_update_state == HL7800_FOTA_PAD)) {
+				process_fw_update_rx(&rx_buf);
+				if (!rx_buf) {
+					break;
+				}
+			}
+#endif
+
+			net_buf_skipcrlf(&rx_buf);
+			if (!rx_buf) {
+				break;
+			}
+
+			frag = NULL;
+			len = net_buf_findcrlf(rx_buf, &frag);
+			if (!frag) {
+				break;
+			}
+
+			out_len = net_buf_linearize(rx_msg, sizeof(rx_msg),
+						    rx_buf, 0, len);
+
+			/* look for matching data handlers */
+			i = -1;
+			for (i = 0; i < ARRAY_SIZE(handlers); i++) {
+				if (ictx.search_no_id_resp) {
+					cmpRes = strncmp(ictx.no_id_resp_cmd,
+							 handlers[i].cmd,
+							 handlers[i].cmd_len);
+				} else {
+					cmpRes =
+						strncmp(rx_msg, handlers[i].cmd,
+							handlers[i].cmd_len);
+				}
+
+				if (cmpRes == 0) {
+					/* found a matching handler */
+
+					/* skip cmd_len */
+					if (!ictx.search_no_id_resp) {
+						rx_buf = net_buf_skip(
+							rx_buf,
+							handlers[i].cmd_len);
+					}
+
+					/* locate next cr/lf */
+					frag = NULL;
+					len = net_buf_findcrlf(rx_buf, &frag);
+					if (!frag) {
+						break;
+					}
+
+					LOG_DBG("HANDLE %s (len:%u)",
+						handlers[i].cmd, len);
+					/* call handler */
+					if (handlers[i].func) {
+						removeLineFromBuf =
+							handlers[i].func(
+								&rx_buf, len);
+					}
+					cmd_handled = true;
+					ictx.search_no_id_resp = false;
+					frag = NULL;
+					/* make sure buf still has data */
+					if (!rx_buf) {
+						break;
+					}
+
+					/* We've handled the current line
+					 * and need to exit the "search for
+					 * handler loop".  Let's skip any
+					 * "extra" data and look for the next
+					 * CR/LF, leaving us ready for the
+					 * next handler search.
+					 */
+					len = net_buf_findcrlf(rx_buf, &frag);
+					break;
+				}
+			}
+
+#ifdef HL7800_LOG_UNHANDLED_RX_MSGS
```

Comment:
It would be better to use `IS_ENABLED()` macro to make sure that the code inside the the block does not bitrot. This is important to the code blocks that are rarely activated. So here (and similar parts of this file), I would suggest this instead:
```
if (IS_ENABLED(HL7800_LOG_UNHANDLED_RX_MSGS) &&
    !cmd_handled && frag && len > 1) {
...
}
```